### ISSUE TITLE
link_to_remove_association not removing persisted records from the DOM when clicked

### ISSUE THREAD

**Issue by diegogonzalesdesouza** [nathanvda/cocoon#567]
I've created a small **Rails (5.2.3)** app as a side-project.
The app consists of a systems where the user can (1) create components, then (2) create compositions, and (3) add available components to the compositions.
The composition and components are related through the composition_components table as follows:
```ruby
class Composition < ApplicationRecord
  has_many :composition_components, inverse_of: :composition, dependent: :destroy
  accepts_nested_attributes_for :composition_components,
                                allow_destroy: true,
                                reject_if: :all_blank
  has_many :components, through: :composition_components
end

class Component < ApplicationRecord
  has_many :composition_components
end

class CompositionComponent < ApplicationRecord
  belongs_to :composition
  belongs_to :component
end
```

This is the cocoon form that is dedicated to create composition_components for a given composition:
```erb
<div class="card">
    <div class="card-body shadow-sm">
      <%= form_for(@composition ? @composition : composition, url: url, method: method) do |f| %>
        <div class="d-flex justify-content-between">
          <div class="flex-grow-1">
            <div class="form-group">
              <%= f.label :name, 'Nome da formulação' %>
              <%= f.text_field :name, class: 'form-control dib h4 flex-grow-1' %>
            </div>
          </div>
        </div>

        <hr class="my-4">

        <div class="d-flex flex-wrap justify-content-between">
          <div class="h4">Composição&nbsp
            <%= link_to_add_association '+ Adicionar Componente', f, :composition_components, partial: 'shared/components/composition_component_fields', class: 'btn btn-primary rounded-pill py-1', data: { association_insertion_node: '.composition_components', association_insertion_method: :prepend } %>
          </div>

          <div class="d-flex w-100 mt-3">
            <div class="col-5 px-0">Componente</div>
            <div class="col-3 px-0">Proporção</div>
            <div class="col-3 px-0">Preço (Kg)</div>
            <div class="col-1 px-0">Excluir</div>
          </div>
          <!-- cocoon for new composition components -->
          <div class="w-100 mb-4 composition_components">
            <%= f.fields_for :composition_components do |cc| %>
              <%= render partial: 'shared/components/composition_component_fields', locals: {f: cc} %>
            <% end %>
          </div>
        </div>

        <%= f.hidden_field :organization_id, value: current_user.organization.id, class: 'd-none' %>

        <%= f.submit('Salvar formulação', class: 'btn btn-secondary mt-3') %>
      <% end %>
    </div>
  </div>
```

And the composition_component_fields partial:
```erb
<div class="nested-fields row w-100 d-flex justify-content-between align-items-center mt-2">
  <div class="col-5">
    <%= f.select(:component_id,
                 options_for_select(current_user.organization.components.collect{ |component| [component.name, component.id] }.prepend(['Selecionar', nil]),
                                    f.object.component_id),
                 {},
                 class: 'form-control select') %>
  </div>

  <div class="col-3">
    <%= f.text_field :weight, value: (f.object.weight if f.object), class: "form-control" %>
  </div>

  <div class="col-3 value">
    <div class="input-group">
      <div class="input-group-prepend">
        <span class="input-group-text currency" id="inputGroupPrepend"></span>
      </div>
      <%= f.text_field :value, value: f.object.component_id.present? ? Component.find(f.object.component_id).value : '', class: "value form-control text-right" %>
    </div>
  </div>

  <div class="w-100 col-1">
    <%= link_to_remove_association f, class: 'link' do %>
      <i class="fas fa-times-circle text-danger h4 pt-1"></i>
    <% end %>
  </div>
</div>
```

The question is that, when I try to remove a new (not persisted) record by clicking on 'link_to_remove_association' it works fine. But, when I try to remove an existing/persisted record, it doesn't remove the element from the DOM.

Tried different workarounds and any of them worked for me. Any clue on what am I missing?

**Comment by nathanvda** [comment]
Just to be clear I understand the behaviour: 

* when you click the link_to_remove_association the (persisted) item is hidden? (but not removed from the DOM because rails needs to know which items to delete from the database, and it will only delete them after post-ing the form)
* it is not hidden? 
* it is hidden but after posting the form to the controller, it re-appears? This could be related to your strong parameters: we need both `id` and `_destroy` to be allowed for removing to work correctly.

**Comment by diegogonzalesdesouza** [comment]
Hey, Nathan, thank you for replying.

Actually it's not even hidden, so I assume the second behaviour you described is the one that is happening. Again, it happens only for persisted items.

**Comment by nathanvda** [comment]
The code is pretty simple: 

~~~
        if ($this.hasClass('dynamic')) {
            node_to_delete.detach();
        } else {
            $this.prev("input[type=hidden]").val("1");
            node_to_delete.hide();
        }
~~~

The `node_to_delete` is correct, or deleting dynamic elements would not work either. So then I guess, something goes wrong when looking for the hidden field `_destroy`. Can you check your javascript console if you get an error there? 

The code is looking for an ancestor on the same level in the DOM, which in your case will not work because you have the link in a separate div. It is looking for the implicitly added `_destroy` field, which is hidden.  The code is supposed to work with mark-up as specified in the documentation. Yours deviates. 

Let me check/think how we can fix this (robustly). The code for the `link_to_remove_association` is simple and specific, but up until now nobody ever requested the same kind of flexibility as we have with the `link_to_add_association`.

**Comment by diegogonzalesdesouza** [comment]
Nathan, my JS is not triggering any error.

Also, I inspected the generated HTML and it seems the code is working fine - even in finding the implicitly added '_destroy'.

I found out that what is preventing the existing element from becoming hidden is that the '.nested-fields' div has an 'd-flex' bootstrap class and that forces display propriety value to 'flex !important '. The '!important' is overriding the 'display: none' that cocoon adds to the excluded/hidden fields.

```erb
<div class="nested-fields row w-100 d-flex justify-content-between align-items-center mt-2" style="display: none;">
  ...
    <input value="1" type="hidden" name="composition[composition_components_attributes][0][_destroy]" id="composition_composition_components_attributes_0__destroy"><a class="link remove_fields existing" href="#">
  ...
</div>
```

The question now is that I'm not sure about why this happens to persisted field and not to new records.

**Comment by nathanvda** [comment]
Ha! Good catch! 

Well, to re-iterate myself, the persisted has to be "hidden" (because it still needs to be in the form with the `_destroy` flag set on save/post), while new records can just be removed from the DOM. 

Could you not simply define a css rule like 

```
.d-flex[style*="display:none"], .d-flex[style*="display: none"] {
    display:none !important;
}
```

Source: https://stackoverflow.com/questions/47514692/bootstrap-4-list-items-that-have-d-flex-class-do-not-respond-to-hide

**Comment by diegogonzalesdesouza** [comment]
Now I got the difference between the behavior of a persisted field and a new one. The solution you mentioned works fine!

Thank you for all the support, Nathan.

**Comment by nathanvda** [comment]
Awesome!
